### ISSUE TITLE
long living TCP connections hang when pod behind a ClusterIP service is restart

### ISSUE THREAD

**Issue by RoadRunnr** [kubernetes/kubernetes#124290]
### What happened?

I've got a service of type ClusterIP with a singe pod backing it. When the pod backing the service is restarted, it gets a new IP and all existing TCP connection just stall.

### What did you expect to happen?

The next packet belonging to existing TCP connections should get a TCP RST.

### How can we reproduce it (as minimally and precisely as possible)?

* use a ClusterIP service with a single backing pod
* establish a long living TCP connection (ssh might work, we used HTTP2)
* kill the pod
* watch the connection stalling

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.29.3
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.27.12+rke2r1
```

</details>

### Cloud provider

<details>
Euqinix Metal
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
PRETTY_NAME="Ubuntu Noble Numbat (development branch)"
NAME="Ubuntu"
VERSION_ID="24.04"
VERSION="24.04 LTS (Noble Numbat)"
VERSION_CODENAME=noble
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=noble
LOGO=ubuntu-logo
$ uname -a
Linux zeus 6.8.0-22-generic #22-Ubuntu SMP PREEMPT_DYNAMIC Thu Apr  4 22:30:32 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by RoadRunnr** [comment]
It is possible that is a dup of https://github.com/projectcalico/calico/issues/7983

**Comment by aojea** [comment]
> The next packet belonging to existing TCP connections should get a TCP RST.

It depends on the implementation of the Service ClusterIP, with kube-proxy  If you have an http2 connection that was not terminated and connected to a ClusterIP that is not longer backed by a Pod, the next packet will be blackholed ... 

If you are using http2 there are two things you should do in your applications:
1. Your Pods must implement graceful termination https://cloud.google.com/blog/products/containers-kubernetes/kubernetes-best-practices-terminating-with-grace
2. enable http2 ping frames to detect idle connections 

/kind support
/remove-kind bug

**Comment by RoadRunnr** [comment]
@aojea thanks for the hint on working around this bug, but this was not the intention of opening the issue.

> connected to a ClusterIP that is not longer backed by a Pod, the next packet will be blackholed ...

That is exactly the problem, the next packet should not be blackholed, it should lead to a reaction, either a TCP RST or a ICMP dest/port unreach.

I could understand if the packet would be blackholed when the pod is completely removed, that would resemble the behavior if a physical host is turned off. However, restarting a host (with the same IP) would result in indications to the peer that the connection has been lost. Restarting a pod should behave in the same way. Now, I understand that the restarted pod gets a new IP, however a pod has much more instrumentation dealing with the surrounding state that a host has. Behaving in graceful and deterministic matter and helping in the cleanup of dangling resource like stale TCP connection should not be a problem.

**Comment by aojea** [comment]
kube-proxy uses NAT to implement the load balancing, so it does not end the connections and does not have a way to  "inform" the endpoints, it works in the middle of the connection so the logic must be implemented in the endpoints or we need to be able to craft packets to terminate the connection.

We discussed this in other places , specifically here https://github.com/kubernetes/kubernetes/issues/108523#issuecomment-1694521909

 TCP RST is not possible as you need to know the TCP sequence to be successful https://datatracker.ietf.org/doc/html/rfc5961#section-3.

I think that ICMP may work https://datatracker.ietf.org/doc/html/rfc5927#section-5 , but is still an invasive solution

It will be nice for someone to come with a PoC to demo how feasible is this and how the implementation will look like, what are the pros and cons and risks ... this seems KEP territory for sure

**Comment by adrianmoisey** [comment]
I tried to reproduce this on GKE and Kubernetes 1.29.
Every time I kill the Pod, it gracefully shutdown the TCP connection for me.
I was testing with a Go web server that held a long HTTP/1.1 connection.

I'll get SSH setup in Pod to give that a try, but do you have more details on how I can successfully reproduce this behaviour?

**Comment by adrianmoisey** [comment]
Ah, I'll try again by catching SIGTERM in my Go web server and ensuring that Go doesn't do a clean shutdown

**Comment by uablrek** [comment]
When a process is killed, even with `kill -9`, the OS (at least Linux) will close the open ports, and the remote side of a TCP connection will get a FIN.
So, the instruction for reproducing this issue must be incomplete.

1. A `Deployment` with replicas=1, and a server inside started that backup a service
2. I connect from an external client to the loadBalancerIP
3. I kill the server process with `kill -9`
4. The POD restarts, but the connection is terminated with a TCP FIN

**Comment by thockin** [comment]
@RoadRunnr we need your help to repro this.  As @adrianmoisey and @aojea said, a clean app shutdown SHOULD be handled cleanly, and all my own testing supports that.

If this is a case of the node going away suddenly, as the SIG discussion theorizes, then we need to ask what we can/should do about it - it's indistinguishable from the other end being slow, which more or less makes it an app-timeout.

**Comment by gjkim42** [comment]
cc

**Comment by thockin** [comment]
Revisiting this 2 weeks in - @RoadRunnr we are looking to you for more input.

**Comment by uablrek** [comment]
I will have another look at this. It may be CNI-plugin specific (https://github.com/kubernetes/kubernetes/issues/124290#issuecomment-2051586556)

/assign

**Comment by uablrek** [comment]
I have reproduced this with [Tero's instructions](https://github.com/tsaarni/kubernetes-tcp-retransmit-retry-timeout/), but _only with Calico in iptables mode_ (the default). Please read the conversation in https://github.com/projectcalico/calico/issues/7983 for a detailed an accurate investigation of the problem.

TL;DR; When the pod is deleted, Calico immediately breaks the route to a pod on another node, but the server runs on for a while, and the FIN sent when it terminates is lost (https://github.com/projectcalico/calico/issues/7983#issuecomment-1768476896).

@RoadRunnr Can you please verify that you are using Calico in iptables mode?

**Comment by uablrek** [comment]
Go sets the TCP keep-alive time to 15s, and I was confused that the client still hangs for 15m. But my conclusion is that TCP keep-alive only is in effect if the connection is idle, but in this case there is a pending write (to the dead server), so the connection is not idle.

**Comment by aojea** [comment]
> Go sets the TCP keep-alive time to 15s, and I was confused that the client still hangs for 15m

https://blog.cloudflare.com/when-tcp-sockets-refuse-to-die

**Comment by RoadRunnr** [comment]
The cluster and setup that had the problem has be deprovisioned, I can therefore not be sure about the exact setup. I do know that it was using calico, but it could have been in nft or iptables mode.

**Comment by uablrek** [comment]
> but it could have been in nft or iptables mode.

No problem, just that it wasn't eBPF-mode (which works).

I close this issue since the problem can't be fixed in K8s, but must be fixed in Calico. Please continue the discussion in https://github.com/projectcalico/calico/issues/7983 (which should be re-opened IMHO).

/close
